Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,267 @@
+  Vol. I*. Springer-Verlag, 1989.
+* [Soare1987] Soare, Robert I. *Recursively Enumerable Sets and Degrees*. Springer-Verlag, 1987.
+* [Gu2015] Gu, Yi-Zhi. *Turing Degrees*. Institute for Advanced Study, 2015.
+
+## Tags
+
+Computability, Oracle, Turing Degrees, Reducibility, Equivalence Relation
+-/
+
+open Primrec Nat.Partrec Part
+
+/--
+The type of partial functions `f` that are recursive in an oracle `g` is the smallest type
+containing the constant zero, the successor, left and right projections, the oracle `g`,
+and is closed under pairing, composition, primitive recursion, and μ-recursion.
+
+Equivalently one can say that `f` is turing reducible to `g` when `f` is recursive in `g`.
+-/
+
+inductive RecursiveIn : (ℕ →. ℕ) → (ℕ →. ℕ) → Prop
+  | zero (g) : RecursiveIn (fun _ => 0) g
+  | succ (g) : RecursiveIn Nat.succ g
+  | left (g) : RecursiveIn (fun n => (Nat.unpair n).1) g
+  | right (g) : RecursiveIn (fun n => (Nat.unpair n).2) g
+  | oracle (g) : RecursiveIn g g
+  | pair (f h g : ℕ →. ℕ) (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun n => (Nat.pair <$> f n <*> h n)) g
+  | comp (f h g : ℕ →. ℕ) (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun n => h n >>= f) g
+  | prec (f h g : ℕ →. ℕ) (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun p =>
+        let (a, n) := Nat.unpair p
+        n.rec (f a) (fun y ih => do
+          let i ← ih
+          h (Nat.pair a (Nat.pair y i)))) g
+  | rfind (f g : ℕ →. ℕ) (hf : RecursiveIn f g) :
+      RecursiveIn (fun a =>
+        Nat.rfind (fun n => (fun m => m = 0) <$> f (Nat.pair a n))) g
+
+
+/--
+`f` is Turing equivalent to `g` if `f` is reducible to `g` and `g` is reducible to `f`.
+-/
+abbrev TuringEquivalent (f g : ℕ →. ℕ) : Prop :=
+  AntisymmRel RecursiveIn f g
+
+/--
+Custom infix notation for `TuringEquivalent`.
+-/
+infix:50 " ≡ᵀ " => TuringEquivalent
+
+/--
+If a function is partial recursive, then it is recursive in every partial function.
+-/
+
+lemma Nat.Partrec.recursiveIn (f : ℕ →. ℕ) (pF : Nat.Partrec f) (g : ℕ →. ℕ) : RecursiveIn f g := by
+  induction pF
+  case zero =>
+    apply RecursiveIn.zero
+  case succ =>
+    apply RecursiveIn.succ
+  case left =>
+    apply RecursiveIn.left
+  case right =>
+    apply RecursiveIn.right
+  case pair f' g' _ _ ih1 ih2 =>
+      apply RecursiveIn.pair f' g' g ih1 ih2
+  case comp f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.comp f' g' g ih1 ih2
+  case prec f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.prec f' g' g ih1 ih2
+  case rfind f' _ ih =>
+    apply RecursiveIn.rfind f' g ih
+
+/--
+If a function is recursive in the constant zero function,
+then it is partial recursive.
+-/
+lemma RecursiveIn.partrec_of_zero (f : ℕ →. ℕ) (fRecInZero : RecursiveIn f fun _ => Part.some 0) :
+  Nat.Partrec f := by
+  generalize h : (fun _ => Part.some 0) = fp at *
+  induction fRecInZero
+  case zero =>
```

Comment:
a styile nit (to apply throughout):
```suggestion
  induction fRecInZero with
  | zero =>
```
etc (this is only valid if you change all the `case` lines)